lib/irb/context.rb

```diff
@@ -490,11 +494,28 @@ def evaluate(line, line_no, exception: nil) # :nodoc:
 
       # Hook command-specific transformation
       command_class = ExtendCommandBundle.load_command(command)
+
+      # If the debugger is activated, we can directly pass debugging command's input
+      if with_debugger && command_class&.ancestors&.include?(ExtendCommand::DebugCommand)
+        throw :IRB_EXIT, original_line
```

Comment:
I think the implementation is a little hacky that `evaluate` does not eval but throw in debug mode, `eval_input`(normally runs IRB's main loop) just return single input.

My idea is doing one of these
- Refactor `each_top_level_statement` and `evaluate`
- Use Fiber to alternately switch to IRB::Debug::UI and IRB's `each_top_level_statement`
- Don't use Debug::UI_Base, store debug session in IRB::Context and implement as ExtendCommand

Refactor like
```ruby
class IRB::Context
  def evaluate(line)
    type, input_or_command = check_input_is_command_or_not(line)
    type == :command ? run_command(input_or_command) : eval(input_or_command)
  end
end
class IRB
  def eval_input
    setup_scanner
    each_top_level_statement{|line|context.evaluate line}
  end
  def debug_read
    setup_scanner
    input = each_top_level_statement{break line}
    context.check_input_is_command_or_not(input)
  end
end
```

Use Fiber like
```ruby
class IRB::Context
  def evaluate(...)
    ...
    if with_debugger
      debugger_ui_fiber.resume original_line
    end
  end
end
class IRB::Debug::UI
  def readline(...)
    debugger_ui_fiber.yield
  end
end
```

Reply:
> `type == :command ? run_command(input_or_command) : eval(input_or_command)`

This is what I want to achieve with #513. Once we can easily distinguish command vs Ruby expression, `Context#evaluate` can really just evaluate the latter and don't care about commands, including debugging commands.

FWIW, I think the current `Context#evaluate` has become `Context#process_line` even before this PR. And IMO the right solution is a big refactor along with #513.

To be clear, I'm by no means saying the current implementation isn't a bit messy. It surely is. But I think it's worth thinking it from some slightly different angles:
- Will the current implementation block future command refactoring?
- Can the implementation become much cleaner with the command refactoring?

I personally think the answers are: No and Yes. Therefore it should be ok to ship it first and refactor later, assume everything behaves as expected.

That being said, it's also worth thinking that if we should implement #513 first, and implement this feature on top of it. This has a few benefits:

- The implementation should be simpler, at least around the `Context#evaluate` part
- We can use this big feature as the attraction for upgrading to v2.0.0

The downsides are:

- We won't be taking such integration into account when implementing 513
- This feature will be shipped and tested much later

@hasumikin @ima1zumi what do you think about this?

Reply:
I feel... regarding this PR, implementing this feature would be good first.

Generally, I think that:
- First, implement a feature if you can make sure it is really necessary and useful
- However, do not leave TOO BIG refactorings in the future
In other words, introducing a feature and refactoring walk in step with each other.

At the same time, @tompng's idea itself looks good to me:

>Refactor each_top_level_statement and evaluate

Reply:
Thanks for the feedback 👍 
While we can't clearly separate command/Ruby expression evaluation atm, I think we can move non-evaluation logic out of `Context#evaluate` first to make things cleaner. So I opened #615 for it.

Reply:
I won't be able to achieve `type == :command ? run_command(input_or_command) : eval(input_or_command)` yet because we can't separate command and expression evaluation atm. But with #618 I will be able to drop `abort`.